### PR TITLE
Update IotHubSasCredential to inherit from TokenCredential

### PR DESCRIPTION
**Pull request by abhipsaMisra**
This PR updates `IotHubSasCredential` type to inherit from `TokenCredential`.
~The service client still defines its own `HttpPipelinePolicy` since `BearerTokenAuthenticationPolicy` requires authentication scopes, which are not available for Iot Hub service.~ Commit 29b55bb5e47ac7a397c83e245dac19a1da62aec0 uses `BearerTokenAuthenticationPolicy`.

Signature for `TokenCredential.GetTokenAsync()` API:
```csharp
public abstract ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken);
```

- `TokenRequestContext ` contains authorization scopes information - currently not relevant for Iot hub service
- `CancellationToken` - the sas token generation is a sync process, a cancellation token isn't really being used here.

### CODE REVIEW

File: sdk/iot/Azure.Iot.Hub.Service/src/Authentication/IotHubSasCredential.cs [hunk 1]
```diff
@@ -104,8 +106,23 @@ private static Uri BuildEndpointUriFromHostName(string hostName)
...
+            }
+
+            var bufferTimeInMilliseconds = (double)RenewalTimeBufferPercentage / 100 * SasTokenTimeToLive.TotalMilliseconds;
+            DateTimeOffset tokenExpiryTimeWithBuffer = _tokenExpiryTime.AddMilliseconds(-bufferTimeInMilliseconds);
+            return DateTimeOffset.UtcNow.CompareTo(tokenExpiryTimeWithBuffer) >= 0;
+        }
+
+        public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
```
Comment (abhipsaMisra) [inline sdk/iot/Azure.Iot.Hub.Service/src/Authentication/IotHubSasCredential.cs:125]: requestContext is unused

File: sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs [hunk 2]
```diff
@@ -14,28 +15,52 @@ namespace Azure.Iot.Hub.Service.Authentication
...
+                ProcessNext(message, pipeline);
+            }
+        }
+
+        private async Task AddHeaders(HttpMessage message, bool async)
+        {
+            var cts = new CancellationTokenSource(_getTokenTimeout);
+
+            AccessToken token = async
+                ? await _iotHubSasCredential.GetTokenAsync(new TokenRequestContext(), cts.Token).ConfigureAwait(false)
```
Comment (abhipsaMisra) [inline sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs:60]: use default constructor for `TokenRequestContext`, other overload requires authorization scopes

File: sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs [hunk 3]
```diff
@@ -14,28 +15,52 @@ namespace Azure.Iot.Hub.Service.Authentication
...
+            }
+            else
+            {
+                ProcessNext(message, pipeline);
+            }
+        }
+
+        private async Task AddHeaders(HttpMessage message, bool async)
+        {
+            var cts = new CancellationTokenSource(_getTokenTimeout);
```
Comment (abhipsaMisra) [inline sdk/iot/Azure.Iot.Hub.Service/src/Authentication/SasTokenAuthenticationPolicy.cs:57]: This cancellation token is used to ensure that we don't wait for > 30secs to retrieve the sas token

File: sdk/iot/Azure.Iot.Hub.Service/src/IotHubServiceClient.cs [hunk 4]
```diff
@@ -21,6 +21,10 @@ public class IotHubServiceClient
         private readonly QueryRestClient _queryRestClient;
         private readonly StatisticsRestClient _statisticsRestClient;
 
+        // IoT Hub service currently does not support OAuth tokens, so they do not have their authorization scopes defined.
+        // This value will need to be correctly populated once OAuth token support is available.
+        private static readonly string[] s_authorizationScopes = new[] { "" };
```
Comment (abhipsaMisra) [inline sdk/iot/Azure.Iot.Hub.Service/src/IotHubServiceClient.cs:26]: This is a weird piece of code that is required for the client to be able to use BearerTokenAuthenticationPolicy

File: sdk/iot/Azure.Iot.Hub.Service/src/Authentication/IotHubSasCredential.cs [hunk 5]
```diff
@@ -122,22 +139,15 @@ public string GetSasToken()
...
-            if (_cachedSasToken == null)
-            {
-                return true;
-            }
-
-            var bufferTimeInMilliseconds = (double)RenewalTimeBufferPercentage / 100 * SasTokenTimeToLive.TotalMilliseconds;
-            DateTimeOffset tokenExpiryTimeWithBuffer = _tokenExpiryTime.AddMilliseconds(-bufferTimeInMilliseconds);
-            return DateTimeOffset.UtcNow.CompareTo(tokenExpiryTimeWithBuffer) >= 0;
+            cancellationToken.ThrowIfCancellationRequested();
+            return GetTokenAsync(requestContext, cancellationToken).Result;
```
Comment (drwill-ms) [inline sdk/iot/Azure.Iot.Hub.Service/src/Authentication/IotHubSasCredential.cs:150]: Should be `.GetAwaiter().GetResult()` instead of `.Result`.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**